src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -292,14 +292,20 @@ public static bool TryDecodeFromUtf16_Vector128(ReadOnlySpan<char> chars, Span<b
                     output = Ssse3.MultiplyAddAdjacent(nibbles,
                         Vector128.Create((short)0x0110).AsSByte()).AsByte();
                 }
-                else
+                else if (AdvSimd.Arm64.IsSupported)
                 {
                     // Workaround for missing MultiplyAddAdjacent on ARM
                     Vector128<short> even = AdvSimd.Arm64.TransposeEven(nibbles, Vector128<byte>.Zero).AsInt16();
                     Vector128<short> odd = AdvSimd.Arm64.TransposeOdd(nibbles, Vector128<byte>.Zero).AsInt16();
                     even = AdvSimd.ShiftLeftLogical(even, 4).AsInt16();
                     output = AdvSimd.AddSaturate(even, odd).AsByte();
                 }
+                else
+                {
+                    // We explicitly recheck each IsSupported query to ensure that the trimmer can see which paths are live/dead
```

Comment:
Could you explain why this added branch is necessary? My expectation is that the entire method should have been trimmed altogether if its preconditions are not met.

Reply:
Some of the checks are "dynamic" and not statically `true`/`false`.

For example, `if (Sse2.IsSupported)` is statically true for `x86/x64` and `if (AdvSimd.Arm64.IsSupported)` is statically `false`, but `if (Ssse3.IsSupported)` is dynamic because it depends on the underlying hardware.

This means that on an `x86/x64` system, the `TryDecodeFromUtf16` will always be preserved in IL. Prior to this PR, the `else` path was therefore being kept for the case that we were on hardware where `Ssse3.IsSupported` was `false` and that rooted `AdvSimd.Arm64` on xarch unnecessarily.

With this change, we instead now allow the `AdvSimd` path to be trimmed out as dead code and the IL will instead keep the `ThrowUnreachableException` which allows things like R2R, the JIT, or AOT compilers to trim it out as dead code (because they will know the target hardware and treat `Ssse3.IsSupported` as constant, then allowing the method to be removed).